File: native/android/app/src/main/java/com/mediconnect/nativeapp/ProfileEnrollment.kt
package com.mediconnect.nativeapp

import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlin.coroutines.coroutineContext

data class OwnProfile(val subject: String, val name: String, val email: String)
data class ProfileDetails(val name: String, val email: String, val specialization: String)
interface ProfileService {
    suspend fun load(identity: Identity): OwnProfile
    suspend fun create(identity: Identity, details: ProfileDetails, policyVersion: String)
}
enum class ProfileStep { CLOSED, CHECKING, MISSING, READY, FAILED }
data class ProfileState(val step: ProfileStep = ProfileStep.CLOSED, val profile: OwnProfile? = null,
                        val accepted: Boolean = false, val busy: Boolean = false)

class ProfileEnrollment(private val service: ProfileService?, private val policyVersion: String,
                        private val scope: CoroutineScope) {
    private val mutable = MutableStateFlow(ProfileState())
    val state = mutable.asStateFlow()
    private var identity: Identity? = null
    private var operation: Job? = null
    private var generation = 0
    fun close() { generation++; operation?.cancel(); identity = null; mutable.value = ProfileState() }
    fun open(value: Identity) {
        close()
        if (service == null || value.role !in setOf(Role.PATIENT, Role.DOCTOR)) return
        identity = value; check()
    }
    fun accept(value: Boolean) {
        if (mutable.value.step == ProfileStep.MISSING && !mutable.value.busy) mutable.value = mutable.value.copy(accepted = value)
    }
    fun check() {
        val identity = identity ?: return
        val service = service ?: return
        if (mutable.value.busy) return
        run {
            mutable.value = ProfileState(ProfileStep.CHECKING, busy = true)
            val profile = readMissing(service, identity)
            mutable.value = if (profile == null) ProfileState(ProfileStep.MISSING) else ProfileState(ProfileStep.READY, profile)
        }
    }
    fun submit(name: String, email: String, specialization: String) {
        val identity = identity ?: return
        val service = service ?: return
        val details = ProfileDetails(name.trim(), email.trim(), specialization.trim())
        if (mutable.value.step != ProfileStep.MISSING || !mutable.value.accepted || mutable.value.busy || policyVersion.isBlank()
            || details.name.isBlank() || details.email.isBlank() || (identity.role == Role.DOCTOR && details.specialization.isBlank())) return
        run {
            val existing = readMissing(service, identity)
            if (existing != null) { mutable.value = ProfileState(ProfileStep.READY, existing); return@run }
            try { service.create(identity, details, policyVersion) }
            catch (failure: ApiFailure) { if (failure.status != 409) throw failure }
            coroutineContext.ensureActive()
            val profile = service.load(identity)
            coroutineContext.ensureActive()
            require(profile.subject == identity.subject)
            mutable.value = ProfileState(ProfileStep.READY, profile)
        }
    }
    private suspend fun readMissing(service: ProfileService, identity: Identity): OwnProfile? {
        val profile = try { service.load(identity) } catch (failure: ApiFailure) {
            if (failure.status != 404) throw failure
            null
        }
        coroutineContext.ensureActive()
        require(profile == null || profile.subject == identity.subject)
        return profile
    }
    private fun run(block: suspend () -> Unit) {
        val current = ++generation
        mutable.value = mutable.value.copy(busy = true)
        operation = scope.launch {
            try { block() }
            catch (cancelled: CancellationException) { throw cancelled }
            catch (_: Exception) { if (current == generation) mutable.value = ProfileState(ProfileStep.FAILED) }
            finally { if (current == generation) mutable.value = mutable.value.copy(busy = false) }
        }
    }
}

File: native/ios/Sources/ProfileEnrollment.swift
import Combine
import Foundation

struct OwnProfile: Equatable { let subject: String; let name: String; let email: String }
struct ProfileDetails { let name: String; let email: String; let specialization: String }
@MainActor protocol ProfileService {
    func load(identity: Identity) async throws -> OwnProfile
    func create(identity: Identity, details: ProfileDetails, policyVersion: String) async throws
}
enum ProfileStep: Hashable { case closed, checking, missing, ready, failed }
struct ProfileState: Equatable {
    var step: ProfileStep = .closed
    var profile: OwnProfile?
    var accepted = false
    var busy = false
}
@MainActor final class ProfileEnrollment: ObservableObject {
    @Published private(set) var state = ProfileState()
    private let service: ProfileService?
    private let policyVersion: String
    private var identity: Identity?
    private var operation: Task<Void, Never>?
    private var generation = 0
    init(service: ProfileService?, policyVersion: String) { self.service = service; self.policyVersion = policyVersion }
    func close() { generation += 1; operation?.cancel(); identity = nil; state = ProfileState() }
    func open(_ identity: Identity) {
        close()
        guard service != nil, identity.role == .patient || identity.role == .doctor else { return }
        self.identity = identity; check()
    }
    func accept(_ value: Bool) { if state.step == .missing && !state.busy { state.accepted = value } }
    func check() {
        guard let identity, let service, !state.busy else { return }
        run { [weak self] in
            self?.state = ProfileState(step: .checking, busy: true)
            let profile = try await Self.readMissing(service, identity: identity)
            self?.state = ProfileState(step: profile == nil ? .missing : .ready, profile: profile)
        }
    }
    func submit(name: String, email: String, specialization: String) {
        let details = ProfileDetails(name: name.trimmingCharacters(in: .whitespacesAndNewlines),
                                     email: email.trimmingCharacters(in: .whitespacesAndNewlines),
                                     specialization: specialization.trimmingCharacters(in: .whitespacesAndNewlines))
        guard let identity, let service, state.step == .missing, state.accepted, !state.busy,
              !policyVersion.isEmpty, !details.name.isEmpty, !details.email.isEmpty,
              identity.role != .doctor || !details.specialization.isEmpty else { return }
        let policyVersion = policyVersion
        run { [weak self] in
            if let existing = try await Self.readMissing(service, identity: identity) {
                self?.state = ProfileState(step: .ready, profile: existing); return
            }
            do { try await service.create(identity: identity, details: details, policyVersion: policyVersion) }
            catch let failure as HTTPFailure { if failure.status != 409 { throw failure } }
            try Task.checkCancellation()
            let profile = try await service.load(identity: identity)
            try Task.checkCancellation()
            guard profile.subject == identity.subject else { throw MobileFailure.response }
            self?.state = ProfileState(step: .ready, profile: profile)
        }
    }
    private static func readMissing(_ service: ProfileService, identity: Identity) async throws -> OwnProfile? {
        let profile: OwnProfile?
        do { profile = try await service.load(identity: identity) }
        catch let failure as HTTPFailure { if failure.status != 404 { throw failure }; profile = nil }
        try Task.checkCancellation()
        guard profile == nil || profile?.subject == identity.subject else { throw MobileFailure.response }
        return profile
    }
    private func run(_ block: @escaping @MainActor () async throws -> Void) {
        generation += 1; let current = generation; state.busy = true
        operation = Task { [weak self] in
            do { try await block() }
            catch is CancellationError { return }
            catch {
                guard let self, generation == current, !Task.isCancelled else { return }
                state = ProfileState(step: .failed)
            }
            if let self, generation == current { state.busy = false }
        }
    }
}
